lib/iris/analysis/_regrid.py

```diff
@@ -1089,20 +971,168 @@ def __call__(self, src):
         )
 
         # Wrap up the data as a Cube.
-        regrid_callback = functools.partial(
-            self._regrid, method=self._method, extrapolation_mode="nan"
+        _regrid_callback = functools.partial(
+            self._regrid,
+            src_x_coord=src_x_coord,
+            src_y_coord=src_y_coord,
+            sample_grid_x=sample_grid_x,
+            sample_grid_y=sample_grid_y,
+            method=self._method,
+            extrapolation_mode="nan",
         )
-        result = self._create_cube(
+
+        def regrid_callback(*args, **kwargs):
+            _data, dims = args
+            return _regrid_callback(_data, *dims, **kwargs)
+
+        result = _create_cube(
             data,
             src,
-            x_dim,
-            y_dim,
-            src_x_coord,
-            src_y_coord,
-            grid_x_coord,
-            grid_y_coord,
-            sample_grid_x,
-            sample_grid_y,
+            [x_dim, y_dim],
+            [grid_x_coord, grid_y_coord],
+            2,
             regrid_callback,
         )
         return result
+
+
+def _create_cube(
+    data, src, src_dims, tgt_coords, num_tgt_dims, regrid_callback
+):
+    r"""
+    Return a new cube for the result of regridding.
+    Returned cube represents the result of regridding the source cube
+    onto the horizontal coordinates (e.g. latitude) of the target cube.
+    All the metadata and coordinates of the result cube are copied from
+    the source cube, with two exceptions:
+        - Horizontal coordinates are copied from the target cube.
+        - Auxiliary coordinates which span the grid dimensions are
+          ignored.
+    Parameters
+    ----------
+    data : array
+        The regridded data as an N-dimensional NumPy array.
+    src : cube
+        The source Cube.
+    src_dims : tuple of int
+        The dimensions of the X and Y coordinate within the source Cube.
+    tgt_coords : tuple of :class:`iris.coords.Coord`\\ 's
+        Either two 1D :class:`iris.coords.DimCoord`\\ 's, two 1D
+        :class:`iris.experimental.ugrid.DimCoord`\\ 's or two ND
+        :class:`iris.coords.AuxCoord`\\ 's representing the new grid's
+        X and Y coordinates.
+    num_tgt_dims : int
+        The number of dimensions that the `tgt_coords` span.
```

Comment:
The idea here is that if you have a pair of 1D coords, these could be describing a 2D rectilinear grid or they could be on the same dimension and describe 1D unstructured data. This is inherited from iris-esmf-regrid where this is necessary. The idea here is that ultimately, iris-esmf-regrid could inherit this function directly from iris.

Reply:
More relevantly, the `UnstructuredNearestNeigbourRegridder` would have to use this when that is refactored. This wasn't refactored as part of this PR since the existing structure is a bit harder to untangle so is worth handling in its own PR.